runtime/common/src/parachains.rs

```diff
@@ -322,6 +429,62 @@ decl_module! {
 			Ok(())
 		}
 
+		/// Provide a proof that some validator has commited a double-vote.
+		pub fn report_double_vote(
+			origin,
+			report: DoubleVoteReport,
+		) -> DispatchResult {
+			ensure_none(origin)?;
+
+			let session_index = <session::Module<T>>::current_index();
+			let validators = <session::Module<T>>::validators();
+			let validator_set_count = validators.len() as u32;
+			let parent_hash = <system::Module<T>>::block_hash(<system::Module<T>>::block_number());
+
+			let authorities = Self::authorities();
+			let offender_idx = match authorities.iter().position(|a| *a == report.identity) {
+				Some(idx) => idx,
+				None => return Err("Not in validator set".into()),
+			};
+
+			let offender = match T::FullIdentificationOf::convert(validators[offender_idx].clone()) {
+				Some(id) => (validators[offender_idx].clone(), id),
+				None => return Err("Failed to convert".into()),
+			};
+
+			let offence = DoubleVoteOffence { session_index, validator_set_count, offender };
+
+			ensure!(
+				report.verify(parent_hash).is_ok(),
+				"Report verification failed.",
+			);
+
+			match (report.first.0, report.second.0) {
+				// If issuing a `Candidate` message on a parachain block, neither a `Valid` or
+				// `Invalid` vote cannot be issued on that parachain block, as the `Candidate`
+				// message is an implicit validity vote.
+				(Statement::Candidate(candidate), Statement::Valid(hash)) |
+				(Statement::Candidate(candidate), Statement::Invalid(hash)) |
+				(Statement::Valid(hash), Statement::Candidate(candidate)) |
+				(Statement::Invalid(hash), Statement::Candidate(candidate))
+				if candidate.hash() == hash => {
+					T::ReportDoubleVote::report_offence(vec![], offence);
```

Comment:
what is the plan to determine `reporters` further on?

Reply:
also - a side PR on Substrate for `report_offence` to return a value if the report should be ignored would be helpful in practice, this is when `pallet-offences` triages and finds that all reports are duplicate.

The reason I want that is so we can reject duplicate offence reports. It should not be possible for a single equivocation to allow huge spam on the chain.

Reply:
Looks like it should be a signed extrinsic and the reporer's Id is `ensure_signed` result?

Reply:
This bears some investigation. My understanding was that the signed extrinsic bears a fee (which we don't want). Maybe setting weight to 0 is the correct approach, though. The other thing you should check before using signed extrinsics is that returning an error means that the extrinsic cannot be included in the block. That's for the same anti-spam reason.